### PR TITLE
refactor: refactor `BeginTransaction` to its own private method

### PR DESCRIPTION
**Pull request by mr-salty**
No functional change. We were doing this in 2 places and there will soon
be additional cases where it is needed.

Part of #4516

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4605)

### CODE REVIEW

File: google/cloud/spanner/internal/connection_impl.cc [hunk 1]
```diff
@@ -301,6 +301,48 @@ Status ConnectionImpl::PrepareSession(SessionHolder& session,
   return Status();
 }
 
+/**
+ * Performs an explicit `BeginTransaction` in cases where that is needed.
+ *
+ * @p func is used to identify the calling function for logging purposes and
```
Comment (coryan) [inline google/cloud/spanner/internal/connection_impl.cc:307]: nit: `@p xxx` just tells Doxygen to format `xxx` with the right font for a parameter. To describe parameters you want `@param` and if you describe one, you must describe all of them.

**Comment by mr-salty** [comment]
___
*[google/cloud/spanner/internal/connection_impl.cc, line 307 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4605#-MCjrD94hyCDFQtpt5qp-r1-307:-MCn8Dmg94kKAGbEv_vt:b-g7ksiv) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bb7ada2e4c56330ef6a9149a10f2eb63e5732384/google/cloud/spanner/internal/connection_impl.cc#L307)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: `@p xxx` just tells Doxygen to format `xxx` with the right font for a parameter. To describe parameters you want `@param` and if you describe one, you must describe all of them.
</blockquote></details>

I was kind of being "informal" since this is a private method, but fleshed it out a little without getting overly verbose.

File: google/cloud/spanner/internal/connection_impl.cc [hunk 2]
```diff
@@ -301,6 +301,53 @@ Status ConnectionImpl::PrepareSession(SessionHolder& session,
...
+    SessionHolder& session,
+    absl::optional<spanner_proto::TransactionSelector>& s, char const* func,
+    bool is_partitioned_dml) {
+  spanner_proto::BeginTransactionRequest begin;
+  begin.set_session(session->session_name());
+  if (is_partitioned_dml) {
+    *begin.mutable_options()->mutable_partitioned_dml() =
+        spanner_proto::TransactionOptions_PartitionedDml();
+  } else {
+    if (!s && !s->has_begin() && !s->has_single_use()) {
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl.cc:326]: `s` should be a pre-condition.  At the very least `!s && s->anything` is UB.

**Comment by mr-salty** [comment]
___
*[google/cloud/spanner/internal/connection_impl.cc, line 326 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4605#-MCjrD94hyCDFQtpt5qp-r2-326:-MCrdKDT1Z9OO8a_o1MJ:byrjl5f) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6102fe35333e9b3c098d0b8886e8033db8622760/google/cloud/spanner/internal/connection_impl.cc#L326)):*
<details><summary><i>Previously, mr-salty (Todd Derr) wrote…</i></summary><blockquote>

I addressed the former by requiring the caller to dereference `s` (obvious in retrospect) and changed the precondition check to a comment.
</blockquote></details>

actually, I guess this is less obvious in retrospect and this method should probably set `s = nullopt` on failure.

File: google/cloud/spanner/internal/connection_impl.cc [hunk 3]
```diff
@@ -301,6 +301,50 @@ Status ConnectionImpl::PrepareSession(SessionHolder& session,
...
+                                        char const* func,
+                                        bool is_partitioned_dml) {
+  spanner_proto::BeginTransactionRequest begin;
+  begin.set_session(session->session_name());
+  if (is_partitioned_dml) {
+    *begin.mutable_options()->mutable_partitioned_dml() =
+        spanner_proto::TransactionOptions_PartitionedDml();
+  } else {
+    *begin.mutable_options() = s.has_begin() ? s.begin() : s.single_use();
+  }
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl.cc:328]: Thought: Pass the options instead of `is_partitioned_dml`.

**Comment by mr-salty** [comment]
dropping in favor of #4624

**Pull request closed without merging**